packages/forms/components/TextInput.tsx

```diff
@@ -0,0 +1,124 @@
+import * as React from "react";
+import { cx } from "emotion";
+import { inputContainer, inputLabel, inputElement } from "../style";
+
+export enum TextInputAppearance {
+  Default = "default",
+  Error = "error",
+  Success = "success"
+}
+
+export interface TextInputProps extends React.HTMLProps<HTMLInputElement> {
+  type: "text" | "number" | "search" | "email" | "password" | "tel" | "url";
+  appearance: TextInputAppearance;
+  inputLabel?: string | React.ReactNode;
+}
+
+export interface TextInputState {
+  focus: boolean;
+}
+
+export class TextInput<T extends TextInputProps> extends React.PureComponent<
+  T,
+  TextInputState
+> {
+  public static defaultProps: Partial<TextInputProps> = {
+    type: "text",
+    appearance: TextInputAppearance.Default
+  };
+
+  constructor(props) {
+    super(props);
+
+    this.inputOnFocus = this.inputOnFocus.bind(this);
+    this.inputOnBlur = this.inputOnBlur.bind(this);
+
+    this.state = {
+      focus: false
+    };
+  }
+
+  public render() {
+    const labelContent = this.getLabelContent();
+    const containerProps: { className?: string } = {};
+    if (this.props.className) {
+      containerProps.className = this.props.className;
+    }
+    return (
+      <div {...containerProps}>
+        {labelContent}
+        {this.getInputContent()}
+      </div>
+    );
+  }
+
+  protected getInputAppearance() {
+    if (this.props.disabled) {
+      return "disabled";
+    }
+    if (this.state.focus) {
+      return `${this.props.appearance}-focus`;
+    }
+    return this.props.appearance;
+  }
+  protected getLabelContent() {
+    if (!this.props.inputLabel) {
+      return null;
+    }
+    const labelProps: { for?: string } = {};
+    if (this.props.id) {
+      labelProps.for = this.props.id;
+    }
+    return (
+      <label className={cx(inputLabel)} {...labelProps}>
+        {this.props.inputLabel}
+      </label>
+    );
+  }
+  protected getInputContent() {
+    return (
+      <span className={cx(inputContainer(this.getInputAppearance()))}>
+        {this.getInputElement()}
+      </span>
+    );
+  }
+  protected getInputElement() {
+    // omit props for container and that we override, otherwise pass through
+    // TextInput props to input element
+    const {
+      type,
+      appearance,
+      inputLabel,
+      onFocus,
+      onBlur,
+      className,
+      ...inputElementProps
+    } = this.props as TextInputProps;
+    return (
+      <input
+        className={cx(inputElement)}
+        type={this.props.type}
+        onFocus={this.inputOnFocus}
+        onBlur={this.inputOnBlur}
+        {...inputElementProps}
+      />
+    );
+  }
+
+  protected inputOnFocus(e: React.FocusEvent<HTMLInputElement>) {
+    this.setState({ focus: true });
```

Comment:
I originally had it using [`:focus-within`](https://css-tricks.com/almanac/selectors/f/focus-within/) since I'm styling the div/span that contains the input, but I was concerned about browser support for that. But I can go back to that, as I'd prefer to not have state here if I can avoid it.

Reply:
[Seems to be problematic on edge](https://caniuse.com/#search=focus-with). I am missing context here, so I don't feel confident to do a judgement if that is an ok trade-off.